### PR TITLE
fix(es/parser): Reject invalid regex flags

### PR DESCRIPTION
**Pull request by g-plane**
**Description:**

Reject unknown regex flags and duplicated regex flags.

Those cases should be considered "Syntax Error": [spec](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#sec-primary-expression-regular-expression-literals-static-semantics-early-errors).

### CODE REVIEW

File: crates/swc_ecma_parser/src/parser/expr.rs [hunk 1]
```diff
@@ -325,11 +325,26 @@ impl<'a, I: Tokens> Parser<I> {
...
+                                let key = match flag {
+                                    'g' | 'i' | 'm' | 's' | 'u' | 'y' => flag,
+                                    _ => '\u{0000}', // special marker for unknown flags
+                                };
+                                map.entry(key).and_modify(|count| *count += 1).or_insert(1);
+                                map
+                            },
+                        );
+                        if flags_count.remove(&'\u{0000}').is_some() {
+                            syntax_error!(self, span, SyntaxError::UnknownRegExpFlags);
```
Comment (kdy1) [inline crates/swc_ecma_parser/src/parser/expr.rs:342]: I think we can use `emit_err` instead?

File: crates/swc_ecma_parser/src/parser/expr.rs [hunk 2]
```diff
@@ -325,11 +325,26 @@ impl<'a, I: Tokens> Parser<I> {
...
+                        }
+                        if let Some((flag, _)) = flags_count.iter().find(|(_, count)| **count > 1) {
+                            syntax_error!(self, span, SyntaxError::DuplicatedRegExpFlags(*flag));
```
Comment (kdy1) [inline crates/swc_ecma_parser/src/parser/expr.rs:345]: Same here

File: crates/swc_ecma_parser/tests/tsc/parserRegularExpressionDivideAmbiguity3.json [hunk 3]
```diff
@@ -52,7 +52,7 @@
                 "ctxt": 0
               },
               "pattern": "regexp",
-              "flags": "a"
+              "flags": "i"
             },
```
Comment (kdy1) [inline crates/swc_ecma_parser/tests/tsc/parserRegularExpressionDivideAmbiguity3.json:56]: This should not be changed

File: crates/swc_ecma_parser/tests/tsc/parserRegularExpressionDivideAmbiguity3.ts [hunk 4]
```diff
@@ -1 +1 @@
-if (1) /regexp/a.foo();
\ No newline at end of file
+if (1) /regexp/i.foo();
```
Comment (kdy1) [inline crates/swc_ecma_parser/tests/tsc/parserRegularExpressionDivideAmbiguity3.ts:1]: Same here

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/swc_ecma_parser/tests/tsc/parserRegularExpressionDivideAmbiguity3.ts (hunk 4 above)
Comment (g-plane) [reply crates/swc_ecma_parser/tests/tsc/parserRegularExpressionDivideAmbiguity3.ts:1]: Test will fail. How should I do?

Comment (kdy1) [reply crates/swc_ecma_parser/tests/tsc/parserRegularExpressionDivideAmbiguity3.ts:1]: I think we need something like https://github.com/swc-project/swc/blob/82069d180463193116ce2d1502934fa824ea8509/crates/swc/tests/tsc.rs#L16-L32 ?

(As these tests are copied from tsc periodically, it's better to ignore)

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_ecma_parser

**Review by kdy1 — APPROVED** [review]

**Pull request merged**